Remotion/Web/Core/Context/HttpContextStorageProviderNetCore.cs

```diff
@@ -0,0 +1,188 @@
+  {
+    private class WrappedHashtable
+    {
+      public WrappedHashtable (HttpContext context, Hashtable data)
+      {
+        Context = context;
+        Data = data;
+      }
+
+      public HttpContext Context { get; }
+      public Hashtable Data { get; }
+    }
+
+    private const string c_dataDictionaryKey = "Rm.DD";
+
+    // Ignore the obsolete warnings for AsyncLocalStorageProvider
+#pragma warning disable 618
+    private readonly AsyncLocalStorageProvider _fallbackProvider = new();
+#pragma warning restore 618
+
+    private readonly ThreadLocal<WeakReference<WrappedHashtable?>> _dataTable = new(() => new WeakReference<WrappedHashtable?>(null));
+
+    public SafeContextBoundary OpenSafeContextBoundary ()
+    {
+      var httpContext = HttpContext.Current;
+      if (httpContext == null)
+      {
+        return _fallbackProvider.OpenSafeContextBoundary();
+      }
+
+      if (TryGetCachedWrappedHashtable(httpContext, out var cachedHashtable))
+      {
+        httpContext.Items.Remove(cachedHashtable.Data);
+        _dataTable.Value!.SetTarget(null!);
+      }
+
+      return new SafeContextBoundary(this, cachedHashtable);
+    }
+
+    public object? GetData (string key)
+    {
+      ArgumentUtility.CheckNotNull("key", key);
+
+      var httpContext = HttpContext.Current;
+      if (httpContext != null)
+      {
+        if (TryGetCachedHashtable(httpContext, out var hashtable))
+        {
+          return hashtable[key];
+        }
+
+        return null;
+      }
+      else
+      {
+        return _fallbackProvider.GetData(key);
+      }
+    }
+
+    public void SetData (string key, object? value)
+    {
+      ArgumentUtility.CheckNotNull("key", key);
+
+      var httpContext = HttpContext.Current;
+      if (httpContext != null)
+      {
+        if (TryGetCachedHashtable(httpContext, out var cachedHashtable))
+        {
+          cachedHashtable[key] = value;
+        }
+        else
+        {
+          var hashtable = SetHashtable(new WrappedHashtable(httpContext, new Hashtable()));
+          hashtable[key] = value;
+        }
+      }
+      else
+      {
+        _fallbackProvider.SetData(key, value);
+      }
+    }
+
+    public void FreeData (string key)
+    {
+      ArgumentUtility.CheckNotNull("key", key);
+
+      var httpContext = HttpContext.Current;
+      if (httpContext != null)
+      {
+        if (TryGetCachedHashtable(httpContext, out var cachedHashtable))
+        {
+          cachedHashtable.Remove(key);
+        }
+      }
+      else
+      {
+        _fallbackProvider.FreeData(key);
+      }
+    }
+
+    void ISafeContextBoundaryStrategy.RestorePreviousSafeContext (object? state)
+    {
+      var httpContext = HttpContext.Current;
+
+      if (state is WrappedHashtable hashtable)
+      {
+        SetHashtable(hashtable);
+      }
+      else if (TryGetCachedHashtable(httpContext, out var cachedHashtable))
+      {
+        httpContext.Items.Remove(cachedHashtable);
```

Comment:
The `Remove` method takes the key not the value so this does not work correctly. We should also write a test to catch that problem.